### PR TITLE
chore: enable BIP44 Accounts other than zero for Ethereum in Native Wallet

### PR DESCRIPTION
**Pull request by asamere**
Enables Bip44 Multi Account support for Ethereum (and evm chains) in HDWallet Native.

### CODE REVIEW

File: packages/hdwallet-native/src/ethereum.ts [hunk 1]
```diff
@@ -57,7 +57,10 @@ export function MixinNativeETHWallet<TBase extends core.Constructor<NativeHDWall
 
     async ethInitializeWallet(masterKey: Isolation.Core.BIP32.Node): Promise<void> {
       const rootNode = await Isolation.Adapters.BIP32.create(masterKey);
-      const isolatedSigner = await rootNode.derivePath(ethers.utils.defaultPath);
+
+      // hard coding this path works fine. need to figure out how to
+      // re-enable supplying a path
```
Comment (asamere) [inline packages/hdwallet-native/src/ethereum.ts:62]: we need to derive the rootNode from a depth 5 path with the account number at index 2. this is due to upstream requirements of ethers.js.

**Review by 0xdef1cafe — APPROVED** [review]

File: packages/hdwallet-native/src/ethereum.ts [hunk 2]
```diff
@@ -53,32 +55,43 @@ export function MixinNativeETHWallet<TBase extends core.Constructor<NativeHDWall
...
 
     async ethGetAddress(msg: core.ETHGetAddress): Promise<string | null> {
-      if (!_.isEqual(msg.addressNList, core.bip32ToAddressNList("m/44'/60'/0'/0/0"))) {
-        throw new Error("path not supported");
+      if (msg.addressNList.length < 5) {
+        throw new Error(`path not supported length: ${msg.addressNList.length}`);
+      }
+
+      // regex sux
+      const refAddressNList = core.bip32ToAddressNList("m/44'/60'/0'/0/0");
```
Comment (asamere) [inline packages/hdwallet-native/src/ethereum.ts:76]: need to make this support chains besides ETH

**Review by 0xdef1cafe — CHANGES_REQUESTED** [review]
disgusting

**Comment by asamere** [comment]
> 

🤮

File: packages/hdwallet-native/src/crypto/isolation/adapters/ethereum.ts [hunk 3]
```diff
@@ -9,18 +10,20 @@ function ethSigFromRecoverableSig(x: SecP256K1.RecoverableSignature): ethers.Sig
...
-    super();
-    this._isolatedKey = isolatedKey;
+  protected constructor(rootNode: Isolation.Adapters.BIP32, provider?: ethers.providers.Provider) {
+    this.rootNode = rootNode;
     this.provider = provider;
   }
 
-  static async create(isolatedKey: SecP256K1.ECDSAKey, provider?: ethers.providers.Provider): Promise<SignerAdapter> {
-    return new SignerAdapter(isolatedKey, provider);
+  static async create(
```
Comment (kaladinlight) [inline packages/hdwallet-native/src/crypto/isolation/adapters/ethereum.ts:22]: we don't really need the static async create anymore if we want to clean that up

File: packages/hdwallet-native/src/crypto/isolation/adapters/ethereum.ts [hunk 4]
```diff
@@ -32,13 +35,15 @@ export class SignerAdapter extends ethers.Signer {
     throw new Error("changing providers on a SignerAdapter is unsupported");
   }
 
-  async getAddress(): Promise<string> {
-    return ethers.utils.computeAddress(SecP256K1.UncompressedPoint.from(await this._isolatedKey.getPublicKey()));
+  async getAddress(addressNList: core.BIP32Path): Promise<string> {
+    const addressNode = await this.rootNode.derivePath(core.addressNListToBIP32(addressNList));
```
Comment (kaladinlight) [inline packages/hdwallet-native/src/crypto/isolation/adapters/ethereum.ts:39]: `addressNode` -> `node` might be a bit better vernacular. No reason these nodes couldn't be derived at any level of the path and just more generic

File: packages/hdwallet-native/src/ethereum.ts [hunk 5]
```diff
@@ -53,32 +54,35 @@ export function MixinNativeETHWallet<TBase extends core.Constructor<NativeHDWall
...
     }
 
     ethWipe() {
       this.#ethSigner = undefined;
     }
 
     async ethGetAddress(msg: core.ETHGetAddress): Promise<string | null> {
-      if (!_.isEqual(msg.addressNList, core.bip32ToAddressNList("m/44'/60'/0'/0/0"))) {
-        throw new Error("path not supported");
+      if (msg.addressNList.length < 5) {
```
Comment (kaladinlight) [inline packages/hdwallet-native/src/ethereum.ts:70]: wondering if we should just omit this check and allow addresses to be derived in non standard ways if desired. we don't really enforce this anywhere else either

Comment (asamere) [reply packages/hdwallet-native/src/ethereum.ts:70]: yep good point

File: packages/hdwallet-native/src/crypto/isolation/adapters/ethereum.ts (hunk 4 above)
Comment (asamere) [reply packages/hdwallet-native/src/crypto/isolation/adapters/ethereum.ts:39]: `rootNode` was definitely the wrong name. I changed it to master which I think reads better and better reflects what the variable is. lmk

Comment (kaladinlight) [reply packages/hdwallet-native/src/crypto/isolation/adapters/ethereum.ts]: parent/child vernacular could be solid as well, parent is passed through (may be root or lower depth) and a child node is derived via the provided path

Comment (asamere) [reply packages/hdwallet-native/src/crypto/isolation/adapters/ethereum.ts]: in the end @pastaghost wins the naming contest with `nodeAdapter`

**Review by 0xdef1cafe — APPROVED** [review]

**Review by gomesalexandre — COMMENTED** [review]
Conceptual stamp, logic looks sane, untested for now

**Review by gomesalexandre — APPROVED** [review]
Tested with linked hdwallet packages, confirmed derivation is working and producing the same address as when adding an Ethereum account with KK

**Pull request merged**